modules/sc-mesh-secure-deployment/src/2_0/features/cbma/tools/utils.py

```diff
@@ -0,0 +1,346 @@
+
+def generate_random_bytes(byte_size=32):
+    return os.urandom(byte_size)
+
+def is_interface_pingable(interface_name, ip_address):
+    # Return true if pingable
+    try:
+        if is_ipv4(ip_address):
+            ping_output = subprocess.check_output(['ping', '-c', '1', '-w', '1', ip_address], stderr=subprocess.STDOUT,universal_newlines=True)
+            return "1 packets transmitted, 1 received" in ping_output
+        elif is_ipv6(ip_address):
+            ping_output = subprocess.check_output(['ping', '-c', '1', '-w', '1', f'{ip_address}%{interface_name}'], stderr=subprocess.STDOUT, universal_newlines=True)
+            return "1 packets transmitted, 1 received" in ping_output
+        else:
+            raise ValueError("Invalid IP address")
+    except subprocess.CalledProcessError as e:
+        return False
+
+def wait_for_interface_to_be_pingable(interface_name, ipv6_address):
+    waiting_message_printed = False
+    while not is_interface_pingable(interface_name, ipv6_address):
+        # Waiting till interface is pingable
+        if not waiting_message_printed:
+            logger.info(f'Waiting for {interface_name} to be reachable..')
+            waiting_message_printed = True
+        time.sleep(1)
+
+def is_interface_up(interface_name):
+    # Check if interface is up
+    try:
+        output = subprocess.check_output(['ifconfig', interface_name], stderr=subprocess.DEVNULL) # Error suppressed as the command throws an error when inyterface is not present
+        return 'inet' in output.decode()
+    except subprocess.CalledProcessError:
+        return False
+
+def wait_for_interface_to_be_up(interface_name):
+    waiting_message_printed = False
+    while not is_interface_up(interface_name):
+        # Waiting till interface is up
+        if not waiting_message_printed:
+            logger.info(f'Waiting for {interface_name} to be up..')
+            waiting_message_printed = True
+        time.sleep(1)
+
+def xor_bytes(byte1, byte2, byte_size=32):
+    # Trim the bytes if they are longer than byte_size
+    if len(byte1) > byte_size:
+        byte1 = byte1[:byte_size]
+    if len(byte2) > byte_size:
+        byte2 = byte2[:byte_size]
+
+    # Pad bytes to the required length
+    if len(byte1) < byte_size:
+        byte1 = byte1.rjust(byte_size, b'\x00')
+    if len(byte2) < byte_size:
+        byte2 = byte2.rjust(byte_size, b'\x00')
+
+    # Return bit-wise XOR of byte1 and byte2
+    return bytes(a ^ b for a, b in zip(byte1, byte2))
+
+def add_interface_to_batman(interface_to_add, batman_interface):
+    # Add interface to batman
+    try:
+        subprocess.run(["batctl", "meshif", batman_interface, "if", "add", interface_to_add], check=True)
+        logger.info(f'Added interface {interface_to_add} to {batman_interface}')
+    except Exception as e:
+        logger.error(f'Error adding interface {interface_to_add} to {batman_interface}: {e}')
+
+def add_interface_to_bridge(interface_to_add, bridge_interface):
```

Comment:
@pentestiing is double-checking but in case it helps to mention in the meantime, please add the Python equivalent of:

`ebtables --tables nat --append OUTPUT --out-interface $MACSEC_INTERFACE --destination ff:ff:ff:ff:ff:ff --jump dnat --to-destination $MACSEC_PEER_MAC_ADDRESS`

This is not mandatory but is good to avoid having batman-adv (at least the upper one) unnecessarily carrying broadcast frames to everyone when it's only the MACsec peer who can decrypt/verify🙂

And of course in a corresponding remove_interface_from_bridge function there should be the cleanup:

`ebtables --tables nat --delete OUTPUT --out-interface $MACSEC_INTERFACE --destination ff:ff:ff:ff:ff:ff --jump dnat --to-destination $MACSEC_PEER_MAC_ADDRESS`